python/tchannel/handler.py

```diff
@@ -0,0 +1,89 @@
+import collections
+
+EndPoint = collections.namedtuple('EndPoint', ['handler', 'opts'])
+
+
+class RequestHandler(object):
+    """Base class for request handlers.
+
+    Usage example:
+        class CustomerReqHandler(RequestHandler):
+            def handle_request(self, context, conn):
+                Add customized request handling
+                logic here
+
+    """
+    def handle_request(self, context, conn):
+        """Handle incoming request
+
+        :param context: context contains received CallRequestMessage
+        :param conn: An incoming TornadoConnection
+        """
+        raise NotImplementedError()
+
+
+class TChannelRequestHandler(RequestHandler):
+    def __init__(self):
+        super(TChannelRequestHandler, self).__init__()
+        self.endpoints = {}
+
+    def handle_request(self, context, conn):
+        """dispatch incoming request to particular endpoint
+
+        :param context: context contains received CallRequestMessage
+        :param conn: An incoming TornadoConnection
+        """
+        request = TChannelRequest(context, conn)
+        endpoint = self._find_endpoint(request.method)
+        if endpoint is not None:
+            response = TChannelResponse(conn)
+            endpoint.handler(request, response, endpoint.opts)
+            response.finish()
+        else:
+            # TODO error handling if endpoint is not found
+            raise NotImplementedError()
+
+    def route(self, rule, **opts):
+        def decorator(handler):
+            self.register_handler(rule, handler, **opts)
+            return handler
+
+        return decorator
+
+    def register_handler(self, rule, handler, **opts):
+        self.endpoints[rule] = EndPoint(handler=handler, opts=opts)
```

Comment:
Unless you're using opts somewhere, don't include it in the `Endpoint`. If and when you add extra options on the `Endpoint`, retrieve them from `opts` and set concrete attributes on the `Endpoint` instead of a formless dict.